### DISCUSSION TITLE
Set environment variable `GPG_TTY` when startting a git process with `process-connection-type` == `pty`

### DISCUSSION THREAD

**Discussion by farblos** [magit/magit#5096] [Ideas]
Emacs Version:

```
GNU Emacs 28.2 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.37, cairo version 1.16.0) of 2023-05-14, modified by Debian
```

Magit version:

```
Magit 20240211.1712 [>= 3.3.0.50-git], Transient 0.5.3, Git 2.39.2, Emacs 28.2, gnu/linux
```

I'd like to use `pinentry-tty` in Emacs as pinentry for GnuPG prompts triggered by (Ma)git (I'll elaborate on the "why" in a follow-up comment).  However, using `pinentry-tty` requires `GPG_TTY` to be set.  Probably Magit could help here?

I guess that for this to work, Magit first would need to acquire a pty and then start the `git` process with `GPG_TTY=<acquired-pty>` added to `process-environment`.  Not sure whether this is technically feasible and you deem it worth the effort, but it could save me another stupid wrapping script.

**Comment by farblos** [comment]
Here is the full story.  A bit lengthy.  More of a show-and-tell, actually.

I'd like to use Ma(git) + GnuPG + ssh + OpenPGP SmartCard (but luckily w/o TRAMP) in the following way:

- Git configured to sign commits and tags with GnuPG and a key on the SmartCard
- Remote repository connected through ssh, where the ssh agent is provided by GnuPG and the private key for ssh authentication is also on the SmartCard
- `pinentry-ncurses` used for all GnuPG operations *outside* of Emacs but some minibuffer prompt for GnuPG operations *triggered by Magit*
- I cannot use `pinentry-emacs` or `allow-emacs-pinentry` since I'm on Debian where any integration between Emacs and GnuPG other than loopback pinentry is verboten
- [I have succeeded in using the loopback pinentry on (Ma)git/Emacs](https://github.com/magit/magit/issues/4076#issuecomment-1950456023), but only for commit signing, and not for the extra indirections that are required once ssh is involved for remote operations.  Besides that, the loopback pinentry prompts are always a bit meager.

Here is what I cobbled up so far:

`~/.gnupg/gpg-agent-conf`:

```magit-git-environment
enable-ssh-support

pinentry-program "/home/directory/bin/pinentry"
```

`~/.ssh/config`

```
Match exec "gpg-connect-agent updatestartuptty /bye"

Host git.sr.ht
  User          git
  # SmartCard key to use for ssh authentication
  IdentityFile ~/.ssh/key.cardno-000f-76ff9c6e.pub
```

`~/bin/git`, a wrapper which sets `GPG_TTY` before calling git:

```
#!/bin/bash

# unwrap the specified executable
unwrap()
{
  local xbn pe oifs

  if [[ $1 = */* ]]; then
    xbn=${1##*/}
  else
    xbn=$1
  fi

  oifs=$IFS; IFS=:; set -f; set -- $PATH; set +f; IFS=$oifs
  for pe in "$@"; do
    if [[ -x "$pe/$xbn" ]] && ! cmp -s "$0" "$pe/$xbn"; then
      printf "$pe/$xbn"
      return 0
    fi
  done

  printf ""
  return 1
}

if ! uwxn=$( unwrap "$0" ); then
  echo "Cannot find unwrapped executable for \"$0\"." 1>&2
  exit 1
fi

if [[ -z ${GPG_TTY-} ]] && tty=$( tty 2>/dev/null ); then
  export GPG_TTY=$tty
fi

exec "$uwxn" "$@"
```

`~/bin/pinentry`, a wrapper to select either `pinentry-curses` or `pinentry-tty`:

```
#!/bin/bash

unwrap()
{
  ... see above
}

if ! uwxn=$( unwrap "$0" ); then
  echo "Cannot find unwrapped executable for \"$0\"." 1>&2
  exit 1
fi

if [[ ${PINENTRY_USER_DATA-} = "INSIDE_MAGIT" ]]; then
  exec pinentry-tty "$@"
else
  exec "$uwxn" "$@"
fi
```

Emacs configuration:

- add regexp `"^Please unlock the card\n\n\\(?:.*\n\\)+PIN: ?$"` to `magit-process-password-prompt-regexps`.  Yes, that matches multiple lines, but at least gives decent information what is being prompted for.
- add string `"PINENTRY_USER_DATA=INSIDE_MAGIT"` to `magit-git-environment`, which triggers usage of `pinentry-tty` in above wrapper script

That fulfills my requirements, but if Magit could help a bit more I'd appreciate that.

**Comment by farblos** [comment]
Yet on the other hand, EPA/EPG does not seem to work well with OpenPGP SmartCards when using loopback pinentry, so it seems that I have to go for `pinentry-emacs`, anyway.  So you might as well just forget about this idea ...

**Reply by tarsius** [reply]
You have a better understanding of the matter than I have. Iff you continue to experiment and find a solution that satisfies you, please let me know.

I also use a openpgp card, but the gui window that pops up when I need to unlock it, does not bother me. I also do not sign over ssh, so I don't have a need to avoid the gui variant.
